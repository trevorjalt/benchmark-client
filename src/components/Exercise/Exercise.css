.ExerciseItem {
    padding: 1vh 0 4vh 0;
}

.ExerciseItem___header {
    display: flex;
    justify-content: center;
    align-items: center;
}

.ExerciseItem__name {
    color: #26547C;
}

.ExerciseItem__name-in-progress {
    color: #26547C;
    margin-right: 1vw;

}

.ExerciseItem__delete, .ExerciseItem__edit, .ExerciseItem__submit {
    display: flex;
    align-items: center;
    margin: 5vh 1vw 5vh 0;

}

.ExerciseItem__delete-icon, .ExerciseItem__edit-icon, .ExerciseItem__submit-icon {
    height: 3.5vh;
}

.ExerciseSetItem__buttons {
    display: flex;
    justify-content: center;
    padding: 4vh 0 4vh 0;
}

.ExerciseSetItem__add {
    display: flex;
    align-items: center; 
}

.ExerciseSetItem__add-icon {
    height: 3.5vh;
    margin-right: 1vw;
}