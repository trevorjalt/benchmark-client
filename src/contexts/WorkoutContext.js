import React, { Component } from 'react'

const WorkoutContext = React.createContext({
    workoutList: [],
    exerciseList: [],
    exerciseSetList: [],
    newWorkoutItem: {},
    error: null,
    isLoggedIn: null,
    clearError: () => {},
    setError: () => {},
    setIsLoggedIn: () => {},
    setWorkoutList: () => {},
    setExerciseList: () => {},
    onDeleteExercise: () => {},
    onDeleteExerciseSet: () => {},
    onDeleteWorkout: () => {},
    onUpdateExercise: () => {},
    onUpdateExerciseSet: () => {},
})
export default WorkoutContext

export class WorkoutProvider extends Component {
    state = {
        workoutList: [],
        exerciseList: [],
        exerciseSetList: [],
        newWorkoutItem: {},
        error: null,
        isLoggedIn: null,
    }

    clearError = () => {
        this.setState({ error: null })
    }

    setError = error => {
        console.error(error)
        this.setState({ error })
    }

    setIsLoggedIn = (value) => {
        this.setState({ isLoggedIn: value })
    }

    setWorkoutList = workoutList => {
        this.setState({ workoutList})
    }

    setExerciseList = exerciseList => {
        this.setState({ exerciseList})
    }

    setExerciseSetList = exerciseSetList  => {
        this.setState({ exerciseSetList })
    }

    onDeleteExercise = (newList) => {
        this.setExerciseList(newList)
    }

    onDeleteExerciseSet = (newList) => {
        this.setExerciseSetList(newList)
    }

    onDeleteWorkout = (newList) => {
        this.setWorkoutList(newList)
    }

    onUpdateExercise = (newList) => {
        this.setExerciseList(newList)
    }

    onUpdateExerciseSet = (newList) => {
        this.setExerciseSetList(newList)
    }

    render() {
        const value = {
            workoutList: this.state.workoutList,
            exerciseList: this.state.exerciseList,
            exerciseSetList: this.state.exerciseSetList,
            newWorkoutItem: this.state.newWorkoutItem,
            error: this.state.error,
            isLoggedIn: this.state.isLoggedIn,
            touched: this.state.touched,
            clearError: this.clearError,
            setError: this.setError,
            setIsLoggedIn: this.setIsLoggedIn,
            setWorkoutList: this.setWorkoutList,
            setExerciseList: this.setExerciseList, 
            setExerciseSetList: this.setExerciseSetList,
            onDeleteExercise: this.onDeleteExercise,
            onDeleteExerciseSet: this.onDeleteExerciseSet,
            onDeleteWorkout: this.onDeleteWorkout,
            onUpdateExercise: this.onUpdateExercise,
            onUpdateExerciseSet: this.onUpdateExerciseSet,         
        }
        return (
            <WorkoutContext.Provider value={value}>
                {this.props.children}
            </WorkoutContext.Provider>
        )
    }
}