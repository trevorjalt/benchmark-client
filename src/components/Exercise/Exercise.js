import React, { Component } from 'react'
import WorkoutContext from '../../contexts/WorkoutContext'
import WorkoutApiService from '../../services/workout-api-service'
import { Button, Section } from '../Utils/Utils'
import ExerciseSet from '../ExerciseSet/ExerciseSet'
import './Exercise.css'


export default class Exercise extends Component {
    static contextType = WorkoutContext

    state = {
        editExerciseName: null,
        addExerciseSet: [],
        updateExerciseName: [],
    }

    componentDidMount() {
        this.context.clearError()
        WorkoutApiService.getExerciseSets()
            .then(this.context.setExerciseSetList)
            .catch(this.context.setError)
    }

    componentDidUpdate(prevProps, prevState) {
        if (prevState.addExerciseSet !== this.state.addExerciseSet) {
            this.context.clearError()
            // WorkoutApiService.getExerciseItem(this.state.newExerciseList.id)
                // .then(this.setState({ newWorkout: !this.state.newWorkout }))
            WorkoutApiService.getExerciseSets()
                .then(this.context.setExerciseSetList)
                .catch(this.context.setError)

        }
    }

    handleClickAddExerciseSet = event => {
        event.preventDefault()
        const { clearError } = this.context
        const { exercise } = this.props
        const newExerciseSet = { set_weight: 0, set_repetition: 0, exercise_id: exercise.id}

        clearError()
        WorkoutApiService.postExerciseSet(newExerciseSet)
            .then(data => this.setState({ addExerciseSet: data }))
        // this.setState({...this.state.addExerciseSet, 
        //     addExerciseSet: {
        //         set_weight: 0,
        //         set_repetition: 0,
        //         exercise_id: exercise.id
        //     }
        // })
        console.log(this.state.addExerciseSet)
        // WorkoutApiService.postExerciseSet(this.state.addExerciseSet)
    }

    handleClickDeleteExercise = event => {
        event.preventDefault()
        const { exerciseList = [], onDeleteExercise, clearError } = this.context
        const { exercise } = this.props
        const newList = exerciseList.filter((item) => item.id !== exercise.id)

        clearError()
        WorkoutApiService.deleteExercise(exercise.id)
            .then(onDeleteExercise(newList))

    }

    handleClickEditExerciseName = () => {
        this.setState({ editExerciseName: !this.state.editExerciseName })
    }

    handleSubmitExerciseName = event => {
        event.preventDefault()
        const { exerciseList = [], clearError, onUpdateExercise } = this.context
        const displayList = exerciseList.map((item, i) => 
            item.id === this.state.updateExerciseName.id 
            ? Object.assign({},item,this.state.updateExerciseName) : item )

        clearError()
        WorkoutApiService.updateExercise(this.state.updateExerciseName)
            .then(onUpdateExercise(displayList))
                // .then(data => this.setState({ updateExerciseName: data }))
            // .then(this.context.setExerciseList)
            .then(this.setState({ editExerciseName: !this.state.editExerciseName }))
    }

    onSelectExerciseNameChange = (event) => {
        const { exercise } = this.props
        this.setState({
            updateExerciseName: {
                id: exercise.id,
                exercise_name: event.target.value,
            }
        })
    }

    renderExerciseSets() {
        const { exerciseSetList = [] } = this.context

        const exerciseSets = exerciseSetList.filter(
            exerciseSet => exerciseSet.exercise_id === this.props.exercise.id)
        return exerciseSets.map(exerciseSet =>
            <ExerciseSet
                key={exerciseSet.id}
                exerciseSet={exerciseSet}
                edit={this.props.edit}
                onRepetitionChange={this.props.onRepetitionChange}
                onWeightChange={this.props.onWeightChange}
                handleClickUpdate={this.props.handleClickUpdate}
            />
        
        )
    }

    renderExerciseState() {
        const { error } = this.context
        const { continueWorkout, exercise, newWorkout } = this.props

        if (this.state.editExerciseName) {
            return (
                <div>
                <form className='EditExerciseNameForm' onChange={this.onSelectExerciseNameChange}>
                        <label htmlFor='ExerciseName__select'></label>
                        <select name='ExerciseName__select' id='ExerciseName__select'
                        // onChange={(event) => this.context.filterSelect(event.target.value)}
                        >
                            <option value=''>Select</option>
                            <option value='Squat'>Squat</option>
                            <option value='Bench'>Bench</option>
                            <option value='Row'>Row</option>
                            <option value='Deadlift'>DeadLift</option>
                            <option value='Military Press'>Military Press</option>
                            <option value='Clean'>Clean</option>
                        </select>

                </form>
                {this.renderSubmitExerciseNameButton()}
                <Section list className='MyExerciseSets'>
                        {error
                        ? <p className='red'>Whoops! There was an error</p>
                        : this.renderExerciseSets()}
                </Section>
                {this.renderAddExerciseSetButton()}
                <span>Exercise Vol: # lbs</span>              
            </div>
                
            ) 
        } else if (continueWorkout || newWorkout) {
            return (
                <div>
                    <div className='ExerciseItem___header'>
                        <h2>{exercise.exercise_name}</h2>
                        {this.renderEditExerciseNameButton()}
                        {this.renderDeleteExerciseButton()}
                    </div>
                    <Section list className='MyExerciseSets'>
                            {error
                            ? <p className='red'>Whoops! There was an error</p>
                            : this.renderExerciseSets()}
                    </Section>
                    {this.renderAddExerciseSetButton()}
                    <span>Exercise Vol: # lbs</span>
                </div>
            )
        } else {
            return (
                <div>
                    <h2>{exercise.exercise_name}</h2>
                    <Section list className='MyExerciseSets'>
                            {error
                            ? <p className='red'>Whoops! There was an error</p>
                            : this.renderExerciseSets()}
                    </Section>
                    <span>Exercise Vol: # lbs</span>              
                </div>
            )            
        }
    }

    renderAddExerciseSetButton() {
        return (
            <Button 
                className='ExerciseSetItem__add' 
                type='button'
                onClick={this.handleClickAddExerciseSet}
            >
                Add Set
            </Button>
        )
    }

    renderDeleteExerciseButton() {
        return (
            <Button 
                className='ExerciseItem__delete' 
                type='button'
                onClick={this.handleClickDeleteExercise}
            >
                ðŸ—‘
            </Button>
        )
    }

    renderEditExerciseNameButton() {
        return (
            <Button 
                className='ExerciseItem__edit' 
                type='button'
                onClick={this.handleClickEditExerciseName}
            >
                âœŽ
            </Button>
        )
    }

    renderSubmitExerciseNameButton() {
        return (
            <Button 
                className='ExerciseItem__submit' 
                type='button'
                onClick={this.handleSubmitExerciseName}
            >
                âœ”
            </Button>
        )
    }

    render() {
        return (
            <div className='ExerciseSet__item'>
                {this.renderExerciseState()}
            </div>
        )
    }
}