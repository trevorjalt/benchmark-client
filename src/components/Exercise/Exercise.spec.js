import React from 'react'
import { WorkoutProvider } from '../../contexts/WorkoutContext'
import Exercise from './Exercise'
import toJson from 'enzyme-to-json'
import { shallow } from 'enzyme'

describe(`Exercise Component`, () => {
    const props = {
        exercise: {
            id: 1,
            exercise_name: 'Limit Breaker',
            workout_id: 1,
        }
    }
    it (`renders without crashing`, () => {
        const wrapper = 
            shallow (
                <WorkoutProvider>
                    <Exercise {...props}/>
                </WorkoutProvider>
            )
        expect(toJson(wrapper)).toMatchSnapshot()
    })
})