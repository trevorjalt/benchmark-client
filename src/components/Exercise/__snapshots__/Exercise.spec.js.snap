// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exercise Component renders without crashing 1`] = `
<ContextProvider
  value={
    Object {
      "clearError": [Function],
      "error": null,
      "exerciseList": Array [],
      "exerciseSetList": Array [],
      "isLoggedIn": null,
      "newWorkoutItem": Object {},
      "onDeleteExercise": [Function],
      "onDeleteExerciseSet": [Function],
      "onDeleteWorkout": [Function],
      "onUpdateExercise": [Function],
      "onUpdateExerciseSet": [Function],
      "setError": [Function],
      "setExerciseList": [Function],
      "setExerciseSetList": [Function],
      "setIsLoggedIn": [Function],
      "setWorkoutList": [Function],
      "touched": undefined,
      "workoutList": Array [],
    }
  }
>
  <Exercise
    exercise={
      Object {
        "exercise_name": "Limit Breaker",
        "id": 1,
        "workout_id": 1,
      }
    }
  />
</ContextProvider>
`;
